### PR TITLE
framework: add sane defaults

### PR DESCRIPTION
**Pull request by kopiczko**

### CODE REVIEW

File: framework/framework.go [hunk 1]
```diff
@@ -94,15 +101,20 @@ func New(config Config) (*Framework, error) {
 		return nil, microerror.Maskf(invalidConfigError, "config.ResourceRouter must not be empty")
 	}
 
+	// Settings.
+	if config.BackOffFactory == nil {
+		return nil, microerror.Maskf(invalidConfigError, "config.BackOffFactory must not be empty")
+	}
+	if config.InitCtxFunc == nil {
+		return nil, microerror.Maskf(invalidConfigError, "config.InitCtxFunc must not be empty")
+	}
```
Comment (kopiczko) [inline framework/framework.go]: Although those are functions to call I can imagine replacing them with few constants to produce values created by these functions. Also they have defaults now. So they are settings?

File: framework/framework.go [hunk 2]
```diff
@@ -21,12 +21,15 @@ import (
 	"github.com/giantswarm/operatorkit/tpr"
 )
 
+const (
+	resourceRetries uint64 = 3
```
Comment (kopiczko) [inline framework/framework.go]: I just took that from kvm-operator.

File: framework/framework.go (hunk 1 above)
Comment (xh3b4sd) [reply framework/framework.go]: I am not sure about this. When you think it is a good thing lets move forward.

File: framework/framework.go [hunk 3]
```diff
@@ -48,31 +51,42 @@ type Config struct {
...
 		Informer:       nil,
-		InitCtxFunc:    nil,
 		Logger:         nil,
 		ResourceRouter: nil,
 		TPR:            nil,
+
+		// Settings.
+		BackOffFactory: func() backoff.BackOff {
+			return backoff.WithMaxTries(backoff.NewExponentialBackOff(), resourceRetries)
+		},
```
Comment (xh3b4sd) [inline framework/framework.go]: Please use the default implementation that is already there. See https://github.com/giantswarm/operatorkit/blob/f21e8578ed653172c0236a1ad7a516409ed125f5/framework/backoff_factory.go. We may want to even remove the file then as well.

File: framework/framework.go (hunk 1 above)
Comment (kopiczko) [reply framework/framework.go]: I'm uncertain as well, but e.g. Caller in micrologger is a setting.

Comment (kopiczko) [reply framework/framework.go]: Motivation here is to be able to change that in all operators globally. Calling it settings is lesser issue for me.

File: framework/framework.go (hunk 3 above)
Comment (kopiczko) [reply framework/framework.go]: Done.

**Comment by kopiczko** [comment]
I don't want to mess with informer. I'd have to add ZeroObjectFactory to Config, and we're removing it soonish. Let's sort it when we move to CRD. Then we'll just pass client and we can create an informer for every K8s object.

**Comment by xh3b4sd** [comment]
I would not mess with zero object factory. My thinking was to set a more proper rate wait duration and resync period maybe.

**Comment by kopiczko** [comment]
> My thinking was to set a more proper rate wait duration and resync period maybe.

I can set those no problem. But I'm not sure what sane defaults are. Currently we have 0 and 1min which sounds ok for me. Maybe we should raise rate wait to 1s to take off some pressure from the API if you ask me.
Do you have some thoughts?

**Comment by xh3b4sd** [comment]
I thought having 1 second and 5 minutes would be good.

**Pull request merged**